### ISSUE TITLE
Certain commands don't work directly after shell commands complete

### ISSUE THREAD

**Issue by rbong** [neovim/neovim#5790]
- `nvim --version`:
```
NVIM 0.1.7
Build type: RelWithDebInfo
Compilation: /usr/bin/cc -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -Wconversion -O2 -g -DDISABLE_LOG -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINC
LUDE_GENERATED_DECLARATIONS -DHAVE_CONFIG_H -D_GNU_SOURCE -I/build/neovim/src/build/config -I/build/neovim/src/neovim-0.1.7/src -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/build/neovim/src/build/sr
c/nvim/auto -I/build/neovim/src/build/include
Compiled by builduser

Optional features included (+) or not (-): +acl   +iconv    +jemalloc +tui
For differences from Vim, see :help vim-differences

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"
```
I have also tried on the newest version.
- Vim (version: ) behaves differently?
```
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Nov 15 2016 18:06:04)
Included patches: 1-86
Compiled by Arch Linux
Huge version without GUI.  Features included (+) or not (-):
+acl             +file_in_path    +mouse_sgr       +tag_old_static
+arabic          +find_in_path    -mouse_sysmouse  -tag_any_white
+autocmd         +float           +mouse_urxvt     +tcl/dyn
-balloon_eval    +folding         +mouse_xterm     +termguicolors
-browse          -footer          +multi_byte      +terminfo
++builtin_terms  +fork()          +multi_lang      +termresponse
+byte_offset     +gettext         -mzscheme        +textobjects
+channel         -hangul_input    +netbeans_intg   +timers
+cindent         +iconv           +num64           +title
-clientserver    +insert_expand   +packages        -toolbar
-clipboard       +job             +path_extra      +user_commands
+cmdline_compl   +jumplist        +perl/dyn        +vertsplit
+cmdline_hist    +keymap          +persistent_undo +virtualedit
+cmdline_info    +lambda          +postscript      +visual
+comments        +langmap         +printer         +visualextra
+conceal         +libcall         +profile         +viminfo
+cryptv          +linebreak       +python/dyn      +vreplace
+cscope          +lispindent      +python3/dyn     +wildignore
+cursorbind      +listcmds        +quickfix        +wildmenu
+cursorshape     +localmap        +reltime         +windows
+dialog_con      +lua/dyn         +rightleft       +writebackup
+diff            +menu            +ruby/dyn        -X11
+digraphs        +mksession       +scrollbind      -xfontset
-dnd             +modify_fname    +signs           -xim
-ebcdic          +mouse           +smartindent     -xpm
+emacs_tags      -mouseshape      +startuptime     -xsmp
+eval            +mouse_dec       +statusline      -xterm_clipboard
+ex_extra        +mouse_gpm       -sun_workshop    -xterm_save
+extra_search    -mouse_jsbterm   +syntax
+farsi           +mouse_netterm   +tag_binary
   system vimrc file: "/etc/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
  fall-back for $VIM: "/usr/share/vim"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H   -D_FORTIFY_SOURCE=2  -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1
Linking: gcc   -L. -Wl,-O1,--sort-common,--as-needed,-z,relro -fstack-protector -rdynamic -Wl,-export-dynamic -Wl,-E -Wl,-rpath,/usr/lib/perl5/core_perl/CORE  -Wl,-O1,--sort-common,--as-needed,-z,relro -L/usr/local/lib -Wl,--as-needed -o vim        -lm -lncurse
s -lelf -lnsl    -lacl -lattr -lgpm -ldl   -Wl,-E -Wl,-rpath,/usr/lib/perl5/core_perl/CORE -Wl,-O1,--sort-common,--as-needed,-z,relro -fstack-protector-strong -L/usr/local/lib  -L/usr/lib/perl5/core_perl/CORE -lperl -lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
 -L/usr/lib -ltclstub8.6 -ldl -lz -lpthread -lieee -lm
```
- Operating system/version:
```
$ uname -r 
4.8.12-2-ARCH
```
- Terminal name/version:
```
rxvt-unicode 9.22-2
```
- `$TERM`:
```
xterm-256color
```
### Actual behaviour
Consider some script that takes input
```
#!/bin/bash
# testscript.sh
echo "enter y/n"
read
```
And consider some functions that use the script
```
fu! RunWithBang()
	" vim gets input and goes to the right window
	" cannot enter input in neovim
	vs
	!testscript.sh
	wincmd l
endfu

fu! RunWithTerminal()
	" does not put you in the right window
	vs
	tabe
	terminal testscript.sh
	wincmd l
endfu

fu! RunWithTermopen()
	" still does not put you in the right window
	vs
	tabe
	let job = termopen('echo y | testscript.sh')
	call jobwait([job])
	wincmd l
endfu
```

The behaviour is in the function comments.
Because the first function does not work in neovim because of #1496, an equivalent version using `terminal` is necessary.
In fact, the issue that spawned #1496, which was that `git add -p` was broken in fugitive, is what spawned this issue. Fugitive added a fix using `:terminal`, but because you cannot wait for the command to finish, it breaks the plugin gitv. Gitv uses fugitive to throw git commands and cannot switch back to its main window afterwards in neovim, causing endless headaches.

### Expected behaviour

I should be able to call commands after terminal programs exit.

Sorry if this is a duplicate. This is difficult to search for.

**Comment by justinmk** [comment]
Fixing  #1496 will implicitly fix this issue. We don't plan to make `:terminal` blocking, but we do plan to restore interactivity to `:!`.

**Comment by justinmk** [comment]
@rbong Instead of using `jobwait()` , could a plugin such as Gitv wait for a result using a `on_stdout` handler passed to `termopen()`?

**Comment by rbong** [comment]
I don't suspect so, since I need to act when the terminal window has fully exited. However, I will test.

**Comment by justinmk** [comment]
Well, there's also `on_exit` handler.

**Comment by rbong** [comment]
I believe that on_exit also failed to give me full control. I will write another example function.

**Comment by justinmk** [comment]
@rbong 

- I don't understand what you expect `wincmd l` to do, because you explicitly did `tabe` there is no window for `wincmd l` to switch to.
- `termopen('echo y | testscript.sh')` won't work unless `testscript.sh` is executable and in your `$PATH`. (Could do `./testscript.sh` instead)

`tabprevious` works instead of `wincmd l`:

```vim
fu! RunWithTermopen()
	" still does not put you in the right window
	vs
	tabe
	let job = termopen('echo y | ./testscript.sh')
	call jobwait([job])
	tabprevious
endfu
```

**Comment by rbong** [comment]
That could be the issue - that Nvim does not close the terminal before it tries to exit and Vim does. If that's the case, I missed that. I'm not sure if that counts as inconsistency between Vim and Neovim - I need to dive back into this issue.

**Comment by rbong** [comment]
@justinmk I have refreshed myself on this issue again.

>I don't understand what you expect wincmd l to do, because you explicitly did tabe there is no window for wincmd l to switch to.

You are right, I tried quite a few different ways to perform window operations after shell scripts complete while putting together the minimal example and somehow I missed this. Thank you.

However, this is still a problem.

>However jobwait() by design blocks Vim until the job exits, which means user can't enter input to testscript.sh.

This was part of my original problem:

>Consider some script that takes input

I wanted to illustrate that I could not find a way to:

1) Start a shell command
2) Get some user input
3) Perform some window operations after the command completes

So while you've illustrated that `1` and `3` are possible, `2` is still missing.

I did not make it clear that I had to put `termopen('echo y | ./testscript.sh')` instead of `termopen('./testscript.sh')` in my examples because `termopen` does not accept user input.

So to summarize:

- `:!` does not accept user input
- `:terminal` does not wait for the command to complete
- `termopen()` does not accept user input
- How can a developer wait for a shell script that accepts user input, as in Vim?

(Note: this message may be just for posterity, I am currently updating Neovim, exploring the issue tracker, and attempting to see if this problem is now fixed)

**Comment by rbong** [comment]
With all of the extra things stripped away, this just needs user input in `:!` (bang) commands.

The relevant discussion is taking place in #1496 

Sorry for overcomplicating this issue.
